src/utils/dashboardModelUtils.ts

```diff
@@ -0,0 +1,142 @@
+  // Helper to remove a widget from contents that get passed in
+
+  // This is composed of 2 Steps:
+
+  /*
+      Step 1:
+        - Find the section that the widget is contained in
+        - Remove it from the section's children
+        - Adjust the relative sizing of the section's remaining children
+        - If the sections children are 0 or 1, we need to do some adjustments
+        - If the above is the case call removeSection to handle this
+
+      Step 2:
+        - Remove the Widget from the Widgets list
+  */
+
+  for(let section in contents.sections) {
+    if(contents.sections[section].children.includes(id)) {
+
+      // Call helper to remove the child from the section and redistribute
+      // sizing amongst the remaining children
+      contents = removeChildAndReDistributeSizing(contents, section, id);
+
+      // IF remaining children is 0 or 1 we have to do some extra stuff
+      // This will be handled by the 'removeSection' function
+      if(contents.sections[section].children.length <= 1) {
+        contents = removeSection(contents, contents.widgets[id].parentId);
+      }
+
+      // Given the assumption that a widget can only be in one section
+      // We break to save some time
+      break;
+    }
+
+  }
+
+  // Finally remove the widget
+  delete contents.widgets[id];
+
+  return contents;
+}
+
+function removeSection(contents: any, id: string) {
+  // Two Cases to Consider:
+
+  /*
+      Case 1: 0 Children in the Section
+        - Remove the Section, recurse on its parent
+
+      Case 2: 1 Child in the Section
+        - We want to remove the section and put the child
+        - in the section's parent section
+  */
+
+  if(id == 'root') {
+    return contents;
+  }
+
+  if(contents.sections[id].children.length == 0) {
+
+    // Call helper to remove the child from the section and redistribute
+    // sizing amongst the remaining children
+    contents = removeChildAndReDistributeSizing(contents, contents.sections[id].parent, id);
+
+    // Recurse if we aren't at the top level
+    if(contents.sections[id].parent != 'root') {
+        contents = removeSection(contents, contents.sections[id].parent)
+    }
+
+    // Remove the section completely
+    delete contents.sections[id];
+
+  } else if(contents.sections[id].children.length == 1) {
+
+    // Take current section's one child and put it into the section's parent section
+    let parentId =contents.sections[id].parent;
+    let currentSectionIndex = contents.sections[parentId].children.indexOf(id);
+    let childId = contents.sections[id].children[0];
+
+    // Replace current section with the child in the section's parent
+    contents.sections[parentId].children[currentSectionIndex] = childId;
+
+    // Checking if the child is a widget or a section
+    if(contents.sections[childId] != null) {
+        contents.sections[childId].parent = parentId;
+    } else {
+        contents.widgets[childId].parentId = parentId;
```

Comment:
oh lol we should probably make these two parameters the same name cause this is silly but out of scope

Reply:
wait actually are you sure they aren't the same name (`parent` and `parentId` for sections and widgets)